File: leetCode/338/338.js
/**
 * @param {number} num
 * @return {number[]}
 */
var recursion = function (num, power, currentSet, result) {
  if (result.length === num + 1) {
    return result
  }
  let elementsToPush = Math.pow(2, power)
  for (let i = 0; i < elementsToPush; i++) {
    // edge case
    if (currentSet[i] === 0) {
      result.push(1)
      continue
    }
    // less than current set => just put my current set value
    if (i < currentSet.length) {
      result.push(currentSet[i])
    } else {
      result.push(currentSet[i - currentSet.length] + 1)
    }
    if (result.length === num + 1) {
      return result
    }
  }
  currentSet = result.slice(result.length - elementsToPush, result.length)
  recursion(num, power+1, currentSet, result)
  return result
}

var countBits = function (num) {
  return recursion(num, 0, [0], [0])
}

console.log(countBits(5))
console.log(countBits(10))

File: leetCode/338.py
class Solution:
    def countBits(self, num: int) -> List[int]:

        def recursion(num: int, power: int, currentSet: List[int], result: List[int]) -> List[int]:

            # array length reach to num + 1 represent can return now
            if len(result) == num + 1:
                return result

            element_to_push = 2**power

            for i in range(element_to_push):
                # when only 1 element
                if i < len(currentSet) and currentSet[i] == 0:
                    result.append(1)
                    continue

                if i < len(currentSet):
                    result.append(currentSet[i])
                else:
                    result.append(currentSet[i - len(currentSet)] + 1)

                # when reached the num
                if len(result) == num + 1:
                    return result

            # update current set
            currentSet = result[len(result) - element_to_push:len(result)]

            recursion(num, power + 1, currentSet, result)

            return result

        return recursion(num, 0, [0], [0])